### ISSUE TITLE
Not clear that JDK immutable factories are more "modern" than Guava immutable collections

### ISSUE THREAD

**Issue by kevinb9n** [gaul/modernizer-maven-plugin#134]
Almost everything we had in Guava that made its way into the JDK came out _better_ in the JDK, and I really appreciate that your tool helps our users get rid of that junk!

But, immutable collections didn't, so unless you define "modern" by "release date" then I'm not sure it really applies to this case.

Couple major advantages of the Guava collections over the newer JDK factories:

*   The set and map have deterministic (insertion-ordered) iteration, instead of randomized order, which produces flakiness (while still being better than undefined ordering, I'll note). 
*   They have a far richer set of construction paths, including builders, and a few other features like `ImmutableSet.asList()`.
*   (the big one) They are public *types* that guarantee immutability, which make for nicer field types and return types than the general-purpose ones like `List`.
    *   This enables static analyzers to disallow modifications at *compile-time*. In spirit, it's nearly as though the modification methods are not even there. (I don't know which tools beyond Error Prone are doing this, but the API makes it _possible_.)
    *   It's not every developer's job to remember when to make a "defensive copy"; just do what the compiler tells you.

The JDK factories have a few advantages as well, especially for code that isn't otherwise using Guava at all. But I don't think a fine-grained discussion of which "wins" overall should be necessary; I think the fact these major trade-offs exist mean that the Guava types are not legacy, and the JDK factories not more modern.

What do you think?

**Comment by Stephan202** [comment]
Just as an extra data point from a company other than Google itself: we have configured the plugin to indeed disable these patterns. Currently our full set of exclusions is as follows:
```xml
<exclusionPatterns>
    <!-- The plugin suggests replacing Guava's
    immutable collection factory methods with
    `java.util.{List,Map,Set}.{copyOf,of}`, but those
    methods don't allow one to convey immutability at
    the type system level. -->
    <exclusionPattern>com/google/common/collect/Immutable(List|Map|Set)\.(copyOf|of):.*</exclusionPattern>
    <!-- The plugin suggests replacing usages of
    Guava's `Iterables` class with
    `java.util.stream.Stream` equivalents, but the
    alternative is often more verbose, requiring
    unnecessary conversions to and from streams. -->
    <exclusionPattern>com/google/common/collect/Iterables\..*</exclusionPattern>
</exclusionPatterns>
```

**Comment by kevinb9n** [comment]
(Now Iterables I'm not sure I'd have gone to bat for. :-))

**Comment by krzyk** [comment]
> The set and map have deterministic (insertion-ordered) iteration, instead of randomized order, which produces flakiness

This is not an advantage. Because of new JDK factories we discovered few bugs in our code that dependent on the order.

Random iteration order was introduced for a reason.

**Comment by krzyk** [comment]
And just another point. 
Most usages of Immutable Collections from guava are just pure and simple creation of given factories, because it is easier to do it e.g. inline during field initialization.

And for those it is pretty valid for modernizer to report them, in other cases one can use excludes.

So besides the system level "Immutable*" I don't see major benefits.

There wasn't a single case that I converted that was troublesome.

**Comment by cpovirk** [comment]
@krzyk Randomized collection ordering is a topic that we've been discussing lately for unrelated reasons. Can you explain what kinds of bugs you fixed when you switched from guaranteed order to randomized order? Traditionally, the bugs we've heard of are from switching from one _non_-guaranteed order to another, as we often see with classes like `HashMap` during JDK upgrades.

**Comment by cpovirk** [comment]
(Maybe you had code that copied a `HashSet` into an `ImmutableSet`, so you were indirectly depending on the non-guaranteed order of `HashSet`? In that case, `Set.copyOf` calls your attention to the problem now, while `ImmutableSet.copyOf` would break only when you upgrade the JDK and see a change in `HashSet` ordering -- or maybe just when your prod code inserts different values that your tests.)

**Comment by gaul** [comment]
tl;dr: I plan to remove these violations before the next release.  I appreciate all the comments on this issue and am documenting my current thinking below.

> I feel I did not present this well at all.

Sorry I was not ignoring this exactly but I hoped that the community would surface some consensus since I lack a strong opinion.  I added the violation in d26cc7c5d6f6bc08ed305397ce149716f1d21c70 in 2014 without too much thought.  Since then the project grew several configuration knobs that remove the need to recommend every possible recommendation and users can add any violations they want.

> * The set and map have deterministic (insertion-ordered) iteration, instead of randomized order, which produces flakiness (while still being better than undefined ordering, I'll note).

I consider this a small win for `Sets` since users should only rely on ordering for `List` types.

> * They have a far richer set of construction paths, including builders, and a few other features like `ImmutableSet.asList()`.

Noted that `Immutable*` have more construction paths although modernizer only warns about the `of` variants.  Although I imagine that most users would prefer to use all or no `Immutable*` for consistency and not a combination of Java and Guava types.

> * (the big one) They are public types that guarantee immutability, which make for nicer field types and return types than the general-purpose ones like `List`.

This is a clever use of deprecation annotations but encouraging use of Guava types in _public_ interfaces is probably a bad idea, especially for libraries that should avoid coupling.  There are decades of debate about the warts of the `Collection` type hierarchy and possible remedies which I am sure you understand better than me.  I will note that `Sets.of` also returns `Set` and that the `Immutable*` variants easily decay to mutable `Collection`s.

My motivation to remove these violations is to reduce the perceived false positive rate for users and thus provide the most utility for the broadest cross-section of users out-of-box.  Thus I plan to remove these violations before the next release.

**Comment by kevinb9n** [comment]
> > I feel I did not present this well at all.
> 
> Sorry I was not ignoring this exactly but

(No, I mean I really think I did present it poorly :-))


> in 2014

jawdrop.


> > * (the big one) They are public types that guarantee immutability, which make for nicer field types and return types than the general-purpose ones like `List`.
> 
> This is a clever use of deprecation annotations but encouraging use of Guava types in _public_ interfaces is probably a bad idea, especially for libraries that should avoid coupling.

Well, the surface area of released libraries is a small minority of collection usages. For the within-Google data point here, we return these types from public methods hundreds of thousands of times, and it works out well. The compile errors aren't even the _best_ part, but since people are generally so hyped about "List shouldn't even have add()", it hits home with them as being the next best thing we can do.

The actual best part goes deeper than that: an immutable implementation means a bad thing won't happen, but an immutable type means you won't have to even *worry* about a bad thing happening, and the difference really helps. I alluded to this re: "defensive copies".


> There are decades of debate about the warts of the `Collection` type hierarchy and possible remedies which I am sure you understand better than me.

Those things will never be remedied in Java itself, so I think what we do is just the next best thing anyone can.

**Comment by gaul** [comment]
> Well, the surface area of released libraries is a small minority of collection usages. For the within-Google data point here, we return these types from public methods hundreds of thousands of times, and it works out well. The compile errors aren't even the _best_ part, but since people are generally so hyped about "List shouldn't even have add()", it hits home with them as being the next best thing we can do.

I will readily admit that I have a limited vantage point since I _exclusively_ work on Java libraries.  I appreciate your feedback since I spend more of my time using C++ these days, both inside Google and otherwise.
